File: atlas/signal-engine/routes/signal.py
"""
signal-engine/routes/signal.py
GET /signal — polled by ATLAS_EA.mq5 every second.
Returns pending signal string or "null".
"""
from fastapi import APIRouter, Request
from fastapi.responses import PlainTextResponse
import structlog

from db.redis_client import get_mode, get_pending_signal

log = structlog.get_logger(__name__)
router = APIRouter()


@router.get("/signal", response_class=PlainTextResponse)
async def get_signal(request: Request) -> str:
    """
    EA polls this endpoint every second when no position is open.

    In AUTO mode:  returns the pending signal immediately.
    In MANUAL mode: returns "null" until trader confirms on dashboard.
                   The confirmed signal is moved to atlas:signal by
                   confirm_manual_signal() when CONFIRM is clicked.

    Returns pipe-delimited signal string:
        "BUY|0.09|1.08000|1.08200|1.08350|1.08500|trade-uuid"
    or "null" if no signal is pending.
    """
    mode = await get_mode()
    sig  = await get_pending_signal()

    if sig:
        log.info("signal_polled", mode=mode, signal=sig[:40])
        return sig

    return "null"

File: atlas/api-server/src/routes/signal.ts
/**
 * api-server/src/routes/signal.ts
 * EA polling endpoints — GET /signal and GET /manage.
 * Plain-text responses (no JSON) — EA parses raw string.
 * riskGuard applied before signal delivery.
 */
import { Router, Request, Response } from "express";
import { riskGuard }                  from "../middleware/riskGuard";
import {
  getMode,
  getPendingSignal,
  getPendingSignalManual,
  getManageCmd,
}                                     from "../services/redisService";

export const signalRouter = Router();

// ─── GET /signal ──────────────────────────────────────────────────────────────
// Polled by ATLAS_EA.mq5 every second when no position is open.
// Returns the pending signal string or "null".

signalRouter.get("/signal", riskGuard, async (req: Request, res: Response) => {
  const mode = await getMode();

  // AUTO mode: return signal immediately for EA execution
  // MANUAL mode: signal sits in manual queue until dashboard confirms
  const sig = mode === "AUTO"
    ? await getPendingSignal()
    : await getPendingSignalManual();

  res.setHeader("Content-Type", "text/plain");
  res.send(sig ?? "null");
});

// ─── GET /manage ──────────────────────────────────────────────────────────────
// Polled by ATLAS_EA.mq5 every second when a position is open.
// Returns the pending management command for the given ticket or "null".

signalRouter.get("/manage", async (req: Request, res: Response) => {
  const ticketStr = req.query["ticket"];
  const ticket    = typeof ticketStr === "string" ? parseInt(ticketStr, 10) : NaN;

  if (isNaN(ticket) || ticket <= 0) {
    res.setHeader("Content-Type", "text/plain");
    res.send("null");
    return;
  }

  const cmd = await getManageCmd(ticket);
  res.setHeader("Content-Type", "text/plain");
  res.send(cmd ?? "null");
});